packages/expo-router/types/expo-router.d.ts

```diff
@@ -0,0 +1,338 @@
+/* eslint-disable @typescript-eslint/no-unused-vars */
+
+declare namespace ExpoRouter {
+  import type { ReactNode } from 'react';
+  import type { TextProps } from 'react-native';
+
+  type StaticRoutes = string;
+  type DynamicRoutes<T extends string> = string;
+  type DynamicRouteTemplate = never;
```

Comment:
> **Note:** These are non-blocking notes, code LGTM 👍
> Just leaving this here in case we come back to this

I chatted with @marklawlor about this in regards to:
- this being a `d.ts` (which we currently need in our approach)
- the potential of isolating this `namespace` to a module explicitly
- applying interface declaration merging to the routes here and separating this from the generated declaration file.

To keep it short, I believe we've got some wiggle room here in how this could be tweaked and implemented differently.
This should give us some options if we run into any situation where we don't like TypeScript errors or have some bad errors when module/declaration issues are introduced in user code.

All of the mentioned changes have some drawbacks and declaration merging would also imply we'd split this declaration here from a declaration that contains the generated routes. So, they're not trivial changes in how this is approached.
But, if users ever see any overly complex TypeScript errors that aren't clear enough, we can try some of the changes above (or others) to tweak it.